packages/@aws-cdk/aws-lambda/lib/function.ts

```diff
@@ -1030,31 +1054,45 @@ Environment variables can be marked for removal when used in Lambda@Edge by sett
     };
   }
 
-  private buildDeadLetterQueue(props: FunctionProps) {
+  private isQueue(deadLetterQueue: sqs.IQueue | sns.ITopic): deadLetterQueue is sqs.IQueue {
+    return (<sqs.IQueue>deadLetterQueue).queueArn !== undefined;
+  }
+
+  private buildDeadLetterQueue(props: FunctionProps): sqs.IQueue | sns.ITopic | undefined {
+    if (!props.deadLetterQueue && !props.deadLetterQueueEnabled && !props.deadLetterTopic) {
+      return undefined;
+    }
     if (props.deadLetterQueue && props.deadLetterQueueEnabled === false) {
       throw Error('deadLetterQueue defined but deadLetterQueueEnabled explicitly set to false');
     }
-
-    if (!props.deadLetterQueue && !props.deadLetterQueueEnabled) {
-      return undefined;
+    if (props.deadLetterTopic && (props.deadLetterQueue || props.deadLetterQueueEnabled !== undefined)) {
+      throw new Error('deadLetterQueue and deadLetterTopic cannot be specified together at the same time');
     }
 
-    const deadLetterQueue = props.deadLetterQueue || new sqs.Queue(this, 'DeadLetterQueue', {
-      retentionPeriod: Duration.days(14),
-    });
-
-    this.addToRolePolicy(new iam.PolicyStatement({
-      actions: ['sqs:SendMessage'],
-      resources: [deadLetterQueue.queueArn],
-    }));
+    let dlqOrDlt: sqs.IQueue | sns.ITopic;
+    if (props.deadLetterTopic) {
+      dlqOrDlt = props.deadLetterTopic;
+      this.addToRolePolicy(new iam.PolicyStatement({
+        actions: ['sns:Publish'],
+        resources: [dlqOrDlt.topicArn],
+      }));
+    } else {
+      dlqOrDlt = props.deadLetterQueue || new sqs.Queue(this, 'DeadLetterQueue', {
+        retentionPeriod: Duration.days(14),
+      });
+      this.addToRolePolicy(new iam.PolicyStatement({
+        actions: ['sqs:SendMessage'],
+        resources: [dlqOrDlt.queueArn],
+      }));
+    }
 
-    return deadLetterQueue;
+    return dlqOrDlt;
```

Comment:
I actually think that keeping this as `deadLetterQueue` makes more sense. I'm thinking of it as a dead letter queue implemented with an sns topic. Lambda [docs](https://aws.amazon.com/about-aws/whats-new/2016/12/aws-lambda-supports-dead-letter-queues/) call it a DLQ exclusively too.

We have to have a `deadLetterTopic` as a property for stability reasons but I think others see it as an implementation detail.

Reply:
I like verbose names. Someone looks at this code and sees that something if up about whatever sits behind this variable. Keeping `deadLetterQueue` and having `sqs.Queue` arounds makes, at least me, to draw just one conclusion and that is that we have a queue in there.

Reply:
Up to you of course but I thought it might make more sense to hint a developers here that `deadLetterQueue` is not exactly a queue.

Reply:
@kaizen3031593 in case notification missed you ;-)

Reply:
@kornicameister I hear you, but I think this distinction can be achieved with some good documentation here. My reasoning is that the data structure representation between `queue` and `topic` does not need to be surfaced via variable name, as they are both `deadLetterQueues`. I agree that it is a bit confusing that a `deadLetterQueue` can be backed by a `topic`, but alas that isn't something i think we can solve here. Say AWS came out with a _third_ backend implementation for `deadLetterQueue`. Would we then have to change the variable name to `dlqOrDltOrDlx`?

The function is also called `buildDeadLetterQueue`... I think adding a documentation line here saying a `dlq` can be backed by `sqs.Queue` and `sns.Topic` is sufficient.

Reply:
@kaizen3031593 variable name changed, thanks for lovely discussion.